[router] Prevent base url from being appended to external links

6TELOIV opened the pull request:
# Why

Fixes #26566

The bug is that `EXPO_BASE_URL` is added to hrefs that are not part of the app's navigation structure. For example, `https://www.example.com/` becomes `/base_urlhttps://www.example.com/`)

# How

Checking that the `href` doesn't include a schema nor authority ensures that we don't add base url to links which should not have it (links like `https://www.example.com/` or `//www.example.com/`).

# Test Plan

Added 3 tests checking base url appending behavior in network relative uris, path uris, and absolute uris. This should cover the URI types that we don't want to append base url to.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

6TELOIV committed:
[router] Add tests for adding baseUrl to Link href param

6TELOIV committed:
[router] Prevent base url from being appended to external urls.

6TELOIV committed:
[router] Add changelog for erroneous base url appending fix

EvanBacon commented on packages/expo-router/src/link/useLinkToPathProps.tsx:
```diff
@@ -52,9 +73,23 @@ export default function useLinkToPathProps({ href, ...options }: UseLinkToPathPr
+    }
+
+    // Append base url only if needed.
+    if (shouldAppendBaseUrl(strippedHref)) {
```
Use `shouldLinkExternally` https://github.com/expo/expo/blob/df05000d67f74aae9c5e03a5d9c5b95807220025/packages/expo-router/src/utils/url.ts#L16

6TELOIV replied:
Changed! Didn't spot that one existing already.

6TELOIV committed:
[router] Remove duplicative code for shouldLinkExternally

6TELOIV commented:
I think I did not run `yarn build`, causing the SDK check to fail. I'm updating the pull request now, my appoligies! This is my first contribution 😛

marklawlor commented:
Same protocol hrefs don't make sense on a native device, so being explicit is better.

But I'm sure there is some edge case where you might need it, so let me discuss this PR https://github.com/expo/expo/pull/31646 with the team first.

6TELOIV commented:
Absolutely; makes sense.

6TELOIV committed:
Only append baseUrl for non-external links

This change was erroneous, and the baseUrl should *only* be prepended when the href is *not* external.

6TELOIV commented:
I realized now that when changing to `shouldLinkExternally` I was appending baseUrl for external links only, when baseUrl should only be appended to non-external links. I forgot to build the package before committing, so I'm doing so now and hopefully will see green lights across the board.

6TELOIV committed:
Merge branch 'expo:main' into main

6TELOIV commented:
I don't have access to a linux env where I can rebuild the package. I will try again later when I'm on my personal device.

6TELOIV committed:
Build expo-router fix for baseUrl appending

6TELOIV commented:
Whew, okay. I think I've got it all working now. Tests were passing on my local, and I properly built the packages. 🤞🤞🤞

6TELOIV commented:
Yay! Green lights 😄
